clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -423,18 +422,28 @@ else if (strategy == OffsetResetStrategy.LATEST)
 
                 nextInLineRecords = parseFetchedData(completedFetch);
             } else {
-                recordsRemaining -= append(drained, nextInLineRecords, recordsRemaining);
+                // check if we've already fetched data for this partition. If so, just return immediately
+                // to avoid the need to recopy the records to a new collection. This case should be rare
+                // because we do not send more than one fetch for a partition at a time.
+
+                TopicPartition partition = nextInLineRecords.partition;
+                if (drained.containsKey(partition))
+                    break;
```

Comment:
I have a slight preference for doing the copy in this case if it's reasonably simple. But I'll leave it to you (maybe I'm being paranoid :)).